drivers/sensor/adi/adxl34x/adxl34x_rtio.c

```diff
@@ -0,0 +1,345 @@
+	 * interrupts.
+	 */
+	if (data->iodev_sqe != NULL) {
+		return 0;
+	}
+	data->iodev_sqe = iodev_sqe;
+
+	/* Enable interrupts on both the MCU and ADXL34x side. */
+	rc = adxl34x_trigger_init(dev);
+	if (rc != 0) {
+		LOG_ERR("Failed to enable the stream");
+		return rc;
+	}
+	rc = adxl34x_trigger_reset(dev);
+	if (rc != 0) {
+		LOG_ERR("Failed to enable the stream");
+		return rc;
+	}
+	return 0;
+#else
+	ARG_UNUSED(dev);
+	ARG_UNUSED(iodev_sqe);
+	return -ENOTSUP;
+#endif /* CONFIG_ADXL34X_TRIGGER */
+}
+
+/**
+ * Collect a single sample of data (x, y and z value) from the sensor
+ *
+ * @param[in] dev Pointer to the sensor device
+ * @param[out] iodev_sqe IO device submission queue entry
+ * @return 0 if successful, negative errno code if failure.
+ */
+static int adxl34x_submit_one_shot(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct adxl34x_dev_data *data = dev->data;
+	const struct adxl34x_cfg *cfg = &data->cfg;
+	int rc;
+
+	const struct sensor_read_config *sensor_config = iodev_sqe->sqe.iodev->data;
+	const struct sensor_chan_spec *const channels = sensor_config->channels;
+	const size_t num_channels = sensor_config->count;
+	const uint32_t min_buf_len = sizeof(struct adxl34x_encoded_data);
+	struct adxl34x_encoded_data *edata;
+	uint8_t *buf;
+	uint32_t buf_len;
+
+	/* Get the buffer for the frame, it may be allocated dynamically by the rtio context. */
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	/* Determine what channels we need to fetch. */
+	for (unsigned int i = 0; i < num_channels; i++) {
+		switch (channels[i].chan_type) {
+		case SENSOR_CHAN_ALL:
+		case SENSOR_CHAN_ACCEL_X:
+		case SENSOR_CHAN_ACCEL_Y:
+		case SENSOR_CHAN_ACCEL_Z:
+		case SENSOR_CHAN_ACCEL_XYZ:
+			break;
+		default:
+			rc = -ENOTSUP;
+			rtio_iodev_sqe_err(iodev_sqe, rc);
+			return rc;
+		}
+	}
+
+	/* Prepare response. */
+	edata = (struct adxl34x_encoded_data *)buf;
+	edata->header.entries = 1;
+	edata->header.range = cfg->data_format.range;
+	edata->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+	edata->header.trigger = (struct adxl34x_int_source){.data_ready = 1};
+
+	rc = adxl34x_rtio_sample_fetch(dev, edata->fifo_data, sizeof(edata->fifo_data));
+	if (rc != 0) {
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	rtio_iodev_sqe_ok(iodev_sqe, 0);
+	return 0;
+}
+
+/**
+ * Collect a single sample or a stream of samples from the sensor
+ *
+ * @param[in] dev Pointer to the sensor device
+ * @param[out] iodev_sqe IO device submission queue entry
+ */
+void adxl34x_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	const struct sensor_read_config *sensor_config = iodev_sqe->sqe.iodev->data;
+	enum pm_device_state pm_state;
+	int rc;
+
+	rc = pm_device_state_get(dev, &pm_state);
```

Comment:
It's really nice we are able to do this automatic power management potentially with read+decode, @JordanYates might be interested or have some comments here on this.